bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java

```diff
@@ -109,24 +109,20 @@ void install(URL url, boolean exclusiveMode) throws IOException {
 			} else {
 				// In this case the platform is up, we should try to do an incremental resolve
 				// TODO consider removing this case because it can cause inconsistent results.
-				refreshPackages(toRefresh.toArray(new Bundle[toRefresh.size()]), manipulatingContext);
-				if (toRefresh.size() > 0) {
-					Bundle[] additionalRefresh = getAdditionalRefresh(prevouslyResolved, toRefresh);
-					if (additionalRefresh.length > 0)
-						refreshPackages(additionalRefresh, manipulatingContext);
+				refreshPackages(toRefresh, manipulatingContext);
+				Collection<Bundle> additionalRefresh = getAdditionalRefresh(prevouslyResolved, toRefresh);
+				if (additionalRefresh.size() > 0) {
```

Comment:
!empty()

Reply:
Is there any advantage? I try to avoid using negations in conditions when possible as its often harder to understand.

Reply:
The size() method can be O(n) in general.

https://stackoverflow.com/questions/27525369/understanding-the-difference-between-collection-isempty-and-collection-size

Reply:
That's quite hypothetical to me... this all uses local standard collection types (in this case `HashSet`) what implement there `isEmpty` as `size == 0`

Reply:
Even tools will complain about such uses (comparing size to 0):

Reply:
Actually the check was not required (anymore) anyways so I removed it now completely.

Beside that, I'm not such a great fan of such "tools" that try to be smart for rare cases and mirco optimizations, I think the only real world example is a [ConcurrentSkipListSet](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListSet.html) where checks for empty/size are meaningless anyways when we assume a high concurrent read/write system...

> Some will state that comparing size() to 0 is "objectively less readable." 

`size() == 0` is equal to `isEmpty()`

but for this case

`size() > 0` is is equal to `!isEmpty()` where one easily can miss the single `!` at the start ;-)

Reply:
> Actually the check was not required (anymore) anyways so I removed it now completely.

Indeed. The code-block where getAdditionalRefresh() is called is only called if `toRefresh` is not empty.

> Beside that, I'm not such a great fan of such "tools" that try to be smart for rare cases and mirco optimizations, I think the only real world example is a [ConcurrentSkipListSet](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListSet.html) where checks for empty/size are meaningless anyways when we assume a high concurrent read/write system...

While I agree that the negation in `!isEmpty()` is not ideal and negations in general tend to make boolean expressions more difficult I would also argue that that writing `not empty` is more meaningful and more like one thinks about the algorithm (at least for me) than `size greater/not equal zero`.
About this specific case I can only tell that I was once bitten by exactly that problem with a filtering list views (like e.g. Guava provides it) where size() must traverse and test all elements in the list, while `isEmpty()` can be a short-circuit.
And in general, I personally like these 'tools' since they often give hints about how one can implement things simpler, more elegant or more performant and thus one can often learn from that and identify code that easaly that needs some rework. And it happened more than once to me that they discovered bugs in code I have just written.
But of course there are also arguable rules/suggestions made by these 'tools', but I digress.